mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala

```diff
@@ -0,0 +1,547 @@
+ * @param link a link function instance
+ */
+private[ml] class Binomial(link: Link = Logit) extends Family(link) {
+
+  override def initialize(y: Double, weight: Double): Double = {
+    val mu = (weight * y + 0.5) / (weight + 1.0)
+    require(mu > 0.0 && mu < 1.0, "The response variable of Binomial family" +
+      s"should be in range (0, 1), but got $mu")
+    mu
+  }
+
+  override def variance(mu: Double): Double = mu * (1.0 - mu)
+
+  override def clean(mu: Double): Double = {
+    if (mu < epsilon) {
+      epsilon
+    } else if (mu > 1.0 - epsilon) {
+      1.0 - epsilon
+    } else {
+      mu
+    }
+  }
+}
+
+private[ml] object Binomial {
+
+  def apply(link: String): Binomial = {
+    link match {
+      case "logit" => new Binomial(Logit)
+      case "probit" => new Binomial(Probit)
+      case "cloglog" => new Binomial(CLogLog)
+    }
+  }
+}
+
+/**
+ * Poisson exponential family distribution.
+ * The default link for the Poisson family is the log link.
+ * @param link a link function instance
+ */
+private[ml] class Poisson(link: Link = Log) extends Family(link) {
+
+  override def initialize(y: Double, weight: Double): Double = {
+    require(y > 0.0, "The response variable of Poisson family " +
+      s"should be positive, but got $y")
+    y
+  }
+
+  override def variance(mu: Double): Double = mu
+
+  override def clean(mu: Double): Double = {
+    if (mu < epsilon) {
+      epsilon
+    } else if (mu.isInfinity) {
+      Double.MaxValue
+    } else {
+      mu
+    }
+  }
+}
+
+private[ml] object Poisson {
+
+  def apply(link: String): Poisson = {
+    link match {
+      case "log" => new Poisson(Log)
+      case "sqrt" => new Poisson(Sqrt)
+      case "identity" => new Poisson(Identity)
+    }
+  }
+}
+
+/**
+ * Gamma exponential family distribution.
+ * The default link for the Gamma family is the inverse link.
+ * @param link a link function instance
+ */
+private[ml] class Gamma(link: Link = Inverse) extends Family(link) {
+
+  override def initialize(y: Double, weight: Double): Double = {
+    require(y > 0.0, "The response variable of Gamma family " +
+      s"should be positive, but got $y")
+    y
+  }
+
+  override def variance(mu: Double): Double = math.pow(mu, 2.0)
+
+  override def clean(mu: Double): Double = {
+    if (mu < epsilon) {
+      epsilon
+    } else if (mu.isInfinity) {
+      Double.MaxValue
+    } else {
+      mu
+    }
+  }
+}
+
+private[ml] object Gamma {
+
+  def apply(link: String): Gamma = {
+    link match {
+      case "inverse" => new Gamma(Inverse)
+      case "identity" => new Gamma(Identity)
+      case "log" => new Gamma(Log)
+    }
+  }
+}
+
+/**
+ * A description of the link function to be used in the model.
+ */
+private[ml] trait Link extends Serializable {
+
+  /** The link function. */
+  def link(mu: Double): Double
+
+  /** Derivative of the link function. */
+  def deriv(mu: Double): Double
+
+  /** The inverse link function. */
+  def unlink(eta: Double): Double
+}
```

Comment:
* We can add `name` to `Link` for easy debugging.
* We can add `Link.fromName(name: String): Link` to handle raw strings.